[receiver/awsfirehosereceiver] Add support for CloudWatch logs

gavincabbage opened the pull request:
**Description:** Add support for CloudWatch logs in JSON format. 

The receiver can now be configured to receiver either `cwmetrics` or `cwlogs` record type for CloudWatch metrics logs, respectively. The former is only valid for metrics receivers and the latter for logs receivers.

Some changes were necessary to the default record type handling to conform to the contracts required by the generated tests and ensure the default config returned by the factory was usable. Previously the default record type was `cwmetrics` but this is not sensible for logs receivers. 

To accommodate this, the default record type is now telemetry-type-specific. That is, for metrics it is `cwmetrics` and for logs it is `cwlogs`. Each receiver makes this determination on its own. Therefore an empty string is now the default returned by the factory and an empty string is also acceptable in the users config. It is no longer required to specify a record type, although it is still supported, so this is not a breaking change.

**Link to tracking Issue:** n/a

**Testing:** Unit tests covering new code as well as manual smoke testing against AWS Firehose and CloudWatch themselves.

**Documentation:** Component README update.

gavincabbage committed:
[receiver/awsfirehosereceiver] Add support for CloudWatch logs

mwear commented on receiver/awsfirehosereceiver/internal/unmarshaler/cwlog/cwlog.go:
```diff
@@ -0,0 +1,17 @@
+
+package cwlog
+
+type cwLog struct {
```
If we want to match the existing `cWMetric` type this should be `cWLog`. I'm not advocating one format over the other, just suggesting uniformity either way.

mwear replied:
If we want to match the existing `cWMetric` type this should be `cWLog`. I'm not advocating one format over the other, just suggesting uniformity either way.

mwear commented on receiver/awsfirehosereceiver/logs_receiver.go:
```diff
@@ -0,0 +1,81 @@
+// The logsConsumer implements the firehoseConsumer
+// to use a metrics consumer and unmarshaler.
+type logsConsumer struct {
+	// consumer passes the translated metrics on to the
```
```suggestion
	// consumer passes the translated logs on to the
```

mwear commented on receiver/awsfirehosereceiver/logs_receiver.go:
```diff
@@ -0,0 +1,81 @@
+)
+
+// The logsConsumer implements the firehoseConsumer
+// to use a metrics consumer and unmarshaler.
```
```suggestion
// to use a logs consumer and unmarshaler.
```

mwear commented on receiver/awsfirehosereceiver/logs_receiver.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+// Consume uses the configured unmarshaler to deserialize the records into a
+// single pmetric.Metrics. If there are common attributes available, then it will
```
```suggestion
// single plog.Logs. If there are common attributes available, then it will
```

mwear reviewed (commented):
I made a first pass through this PR. It very closely mirrors what currently exists for metrics. I would like to have eyes from someone with more expertise on this. @Aneurysm9 can you possibly take a look or recommend someone who can?

gavincabbage committed:
[awsfirehosereceiver/cwlogs] Refactor default record type logic; PR feedback

gavincabbage committed:
[awsfirehosereceiver/cwlogs] Correct comments

gavincabbage commented:
Thanks for the feedback @mwear! I believe addressed your comments and the test failures. 

I did a bit of refactoring around the default record type logic to conform to the contracts required by the generated tests and have updated the description accordingly.

gavincabbage committed:
[awsfirehosereceiver/cwlogs] Remove unused constants

gavincabbage committed:
[awsfirehosereceiver/cwlogs] Remove unconventional common attributes

gavincabbage commented:
There are a couple CI failures still but it is unclear to me what the problem is.

jackgopack4 commented:
> There are a couple CI failures still but it is unclear to me what the problem is.

so for `build-and-test/checks` the test output says "run `make generate` and commit the changes"

I tried that on my local machine and it generated some changes to `receiver/awsfirehosereceiver/generated_component_test.go` and `receiver/awsfirehosereceiver/generated_package_test.go`

Might not solve the cross-compile stage but maybe would check off one of the failing tests

jackgopack4 commented on receiver/awsfirehosereceiver/generated_component_test.go:
```diff
@@ -46,21 +53,21 @@ func TestComponentLifecycle(t *testing.T) {
-	for _, tt := range tests {
-		t.Run(tt.name+"-shutdown", func(t *testing.T) {
-			c, err := tt.createFn(context.Background(), receivertest.NewNopSettings(), cfg)
+	for _, test := range tests {
```
```suggestion
	for _, tt := range tests {
```

jackgopack4 commented on receiver/awsfirehosereceiver/generated_component_test.go:
```diff
@@ -46,21 +53,21 @@ func TestComponentLifecycle(t *testing.T) {
-		t.Run(tt.name+"-shutdown", func(t *testing.T) {
-			c, err := tt.createFn(context.Background(), receivertest.NewNopSettings(), cfg)
+	for _, test := range tests {
+		t.Run(test.name+"-shutdown", func(t *testing.T) {
```
```suggestion
		t.Run(tt.name+"-shutdown", func(t *testing.T) {
```

jackgopack4 commented on receiver/awsfirehosereceiver/generated_component_test.go:
```diff
@@ -46,21 +53,21 @@ func TestComponentLifecycle(t *testing.T) {
-			c, err := tt.createFn(context.Background(), receivertest.NewNopSettings(), cfg)
+	for _, test := range tests {
+		t.Run(test.name+"-shutdown", func(t *testing.T) {
+			c, err := test.createFn(context.Background(), receivertest.NewNopSettings(), cfg)
```
```suggestion
			c, err := tt.createFn(context.Background(), receivertest.NewNopSettings(), cfg)
```

jackgopack4 commented on receiver/awsfirehosereceiver/generated_component_test.go:
```diff
@@ -46,21 +53,21 @@ func TestComponentLifecycle(t *testing.T) {
 		})
-		t.Run(tt.name+"-lifecycle", func(t *testing.T) {
-			firstRcvr, err := tt.createFn(context.Background(), receivertest.NewNopSettings(), cfg)
+		t.Run(test.name+"-lifecycle", func(t *testing.T) {
```
```suggestion
		t.Run(tt.name+"-lifecycle", func(t *testing.T) {
```

jackgopack4 commented on receiver/awsfirehosereceiver/generated_component_test.go:
```diff
@@ -46,21 +53,21 @@ func TestComponentLifecycle(t *testing.T) {
-		t.Run(tt.name+"-lifecycle", func(t *testing.T) {
-			firstRcvr, err := tt.createFn(context.Background(), receivertest.NewNopSettings(), cfg)
+		t.Run(test.name+"-lifecycle", func(t *testing.T) {
+			firstRcvr, err := test.createFn(context.Background(), receivertest.NewNopSettings(), cfg)
```
```suggestion
			firstRcvr, err := tt.createFn(context.Background(), receivertest.NewNopSettings(), cfg)
```

jackgopack4 commented on receiver/awsfirehosereceiver/generated_component_test.go:
```diff
@@ -46,21 +53,21 @@ func TestComponentLifecycle(t *testing.T) {
 			require.NoError(t, firstRcvr.Start(context.Background(), host))
 			require.NoError(t, firstRcvr.Shutdown(context.Background()))
-			secondRcvr, err := tt.createFn(context.Background(), receivertest.NewNopSettings(), cfg)
+			secondRcvr, err := test.createFn(context.Background(), receivertest.NewNopSettings(), cfg)
```
```suggestion
			secondRcvr, err := tt.createFn(context.Background(), receivertest.NewNopSettings(), cfg)
```

jackgopack4 commented on receiver/awsfirehosereceiver/generated_package_test.go:
```diff
@@ -3,9 +3,8 @@
 import (
-	"testing"
-
 	"go.uber.org/goleak"
```
```suggestion
	"testing"

```

jackgopack4 commented on receiver/awsfirehosereceiver/generated_package_test.go:
```diff
@@ -3,9 +3,8 @@
-	"testing"
-
 	"go.uber.org/goleak"
+	"testing"
```
```suggestion
	"go.uber.org/goleak"
```

jackgopack4 reviewed (changes requested):
left suggestions for what the `make generate` command outputted on my machine. Looks like that might help you pass the build-and-test/check stage

gavincabbage committed:
[awsfirehosereceiver/cwlogs] Run make generate

gavincabbage commented on receiver/awsfirehosereceiver/logs_receiver.go:
```diff
@@ -0,0 +1,81 @@
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsfirehosereceiver/internal/unmarshaler"
+)
+
+// The logsConsumer implements the firehoseConsumer
```
Omitted `The/A` entirely to follow the usual godoc approach of starting with the subject name (not that it matters much for an unexported type).

gavincabbage committed:
[awsfirehosereceiver/cwlogs] Doc fixes; add additional common attributes

gavincabbage commented on receiver/awsfirehosereceiver/internal/unmarshaler/cwlog/unmarshaler.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+// isValid validates that the cWMetric has been unmarshalled correctly.
+func (u Unmarshaler) isValid(log cwLog) bool {
```
I don't think so. All these attributes are AWS IDs so they're arbitrary strings. I based this validation on what I saw the metric path doing. That path adds more attributes than I have here because metric payloads have more information (e.g. region, which is not included in logs payloads) but otherwise followed the established pattern.

gavincabbage commented on receiver/awsfirehosereceiver/internal/unmarshaler/cwlog/unmarshaler.go:
```diff
@@ -0,0 +1,107 @@
+		}
+		// Multiple logs in each record separated by newline character
+		for datumIndex, datum := range bytes.Split(record, []byte(recordDelimiter)) {
+			if len(datum) > 0 {
```
I think it is unlikely as this is the wrapper coming from CloudWatch and not the user's log payload itself (which could be anything). 

Out of curiosity though, how would you check for valid JSON besides trying to unmarshal it?

jackgopack4 replied:
You're right as long as unmarshal has enough safeguards it's probably fine.

gavincabbage commented:
Thanks @jackgopack4! Sorry I missed that bit about `make generate` 🤦 

Besides the cross compile issue, I think I've addressed everything you called out, with either a change or a question/response.

jackgopack4 reviewed (approved):
lgtm, thanks for humoring my suggestions

gavincabbage commented on receiver/awsfirehosereceiver/testdata/cwlogs_config.yaml:
```diff
@@ -0,0 +1,7 @@
+awsfirehose:
+  endpoint: 0.0.0.0:4433
```
Sorry I missed this question. 

This receiver works just fine with `localhost` with the caveat that Firehose needs to be able to reach it (Firehose also requires an HTTPS endpoint, I believe). In my testing I've used `ngrok` to facilitate this.
